devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy

```diff
@@ -90,33 +90,66 @@ def parse(String description) {
             List<Map> descMaps = collectAttributes(descMap)
             def liftmap = descMaps.find { it.attrInt == ATTRIBUTE_POSITION_LIFT }
             if (liftmap) {
-                state.level = Integer.parseInt(liftmap.value, 16)
                 if (liftmap.value == "64") { //open
-                    sendEvent(name: "windowShade", value: "open")
                     sendEvent(name: "level", value: "100")
+                    levelEventHandler(100)
                 } else if (liftmap.value == "00") { //closed
-                    sendEvent(name: "windowShade", value: "closed")
                     sendEvent(name: "level", value: "0")
+                    levelEventHandler(0)
                 } else {
-                    sendEvent(name: "windowShade", value: "partially open")
                     sendEvent(name: "level", value: zigbee.convertHexToInt(liftmap.value))
+                    levelEventHandler(zigbee.convertHexToInt(liftmap.value))
                 }
+                state.level = Integer.parseInt(liftmap.value, 16)
             }
         } else if (!supportsLiftPercentage() && descMap?.clusterInt == zigbee.LEVEL_CONTROL_CLUSTER && descMap.value) {
             def valueInt = Math.round((zigbee.convertHexToInt(descMap.value)) / 255 * 100)
-            state.level = valueInt
-            if (0 == valueInt) {
-                sendEvent(name: "windowShade", value: "closed")
-            } else if (100 == valueInt) {
-                sendEvent(name: "windowShade", value: "open")
-            } else {
-                sendEvent(name: "windowShade", value: "partially open")
-            }
+
+            levelEventHandler(valueInt)
             sendEvent([name: "level", value: valueInt])
+            state.level = valueInt
         }
     }
 }
 
+def levelEventHandler(currentLevel) {
+    log.debug "levelEventHandler:- ${currentLevel}"
+    def lastLevel = state.level
+    if (currentLevel == lastLevel) { //Ignore invalid reports
+        return
+    }
+    if (currentLevel == 0) {
+        sendEvent(name: "windowShade", value: "closed")
+        return
+    } else if (currentLevel == 100) {
+        sendEvent(name: "windowShade", value: "open")
+        return
+    }
+    if (lastLevel != null) {
+        if (lastLevel < currentLevel) {
+            sendEvent([name:"windowShade", value: "opening"])
+        } else if (lastLevel > currentLevel) {
+            sendEvent([name:"windowShade", value: "closeing"])
+        }
+    }
+
+    if (device.getDataValue("manufacturer") == "REXENSE") {
+        runIn(3, "updateFinalState", [overwrite:true]) //Dooya should waitting more time to get report
+    } else {
+        runIn(1, "updateFinalState", [overwrite:true])
+    }
+}
+
+def updateFinalState() {
+    def level = state.level
+    log.debug "updateFinalState:- ${level}"
+    if (0 == level||100 == level) {
```

Comment:
you should just invert this conditional so you don't have to have this extra unnecessary `return` statement.